torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -579,3 +579,50 @@ TORCH_LIBRARY(_c10d_functional_autograd, m) {
           c10::DispatchKey::Autograd, ::all_gather_into_tensor_autograd),
       {at::Tag::pt2_compliant_tag});
 }
+
+namespace {
+// DTensor related comm operations, sharing code with functional collective for
+// now
+at::Tensor shard_dim_alltoall(
+    const at::Tensor& input,
+    int64_t gather_dim,
+    int64_t shard_dim,
+    std::string group_name) {
+  auto group = c10d::resolve_process_group(group_name);
+  auto group_size = group->getSize();
+  std::vector<int64_t> output_sizes = input.sizes().vec();
+  if (output_sizes[shard_dim] % group_size != 0) {
+    LOG(WARNING) << "The first dimension of the shard_dim_alltoall input ("
+                 << output_sizes[shard_dim]
+                 << ") is not divisible by the group size (" << group_size
+                 << ").";
+  }
+  output_sizes[shard_dim] = output_sizes[shard_dim] / group_size;
+  std::vector<at::Tensor> inputs;
+  auto length = output_sizes[shard_dim];
+  for (int i = 0; i < group_size; i++) {
+    inputs.push_back(input.narrow(shard_dim, i * length, length).contiguous());
```

Comment:
just FYI (I got tagged on internal since I'm the current PT2 operator oncall) - this op is marked as pt2_compliant, but it doesn't look like it's implemented in a way that will work with dynamic shapes. Let me know if you want some pointers for fixing it!